Support pre-set and post-set parameter callbacks in addition to on-set-parameter-callback. (#1947)
ros2/rclcpp: rclcpp/doc/proposed_node_parameter_callbacks.md

# Proposed node parameters callback Design

## Introduction:

>  The original requirement came in **gazeb_ros_pkgs** for setting individual wheel slip params based on global wheel slip value [link to original issue](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1365). The main idea was to set some parameters using ```set_parameter``` API once the a given param was set successfully. The requirement let to the discussion of supporting registered callbacks once the parameters are set and have been requested by users before. Some related discussion here [#609](https://github.com/ros2/rclcpp/issues/609) [#1789](https://github.com/ros2/rclcpp/pull/1789).

Reviewer: style nitpick: please use one sentence per line to help keep future diffs easier to read (see our [style guide](https://docs.ros.org/en/rolling/Contributing/Code-Style-Language-Versions.html#markdown-restructured-text-docblocks))

> In the current Node API, the ```add_on_set_parameters_callback``` is used for doing any validation for parameters before setting the parameters successfully. Once the parameters are validated ```ParameterEventHandler``` object is used to publish any changes to node parameters on **/parameter_events** topic, which can be subscribed by any node to see the changed parameter. Note that the **/parameter_events** is a topic for all nodes on the network, and we have to rely on executors to process those.

Reviewer: I think only one set of backticks are necessary for inline formatting in markdown,
```suggestion
In the current Node API, the `add_on_set_parameters_callback` is used for doing any validation for parameters before setting the parameters successfully. Once the parameters are validated `ParameterEventHandler` object is used to publish any changes to node parameters on **/parameter_events** topic, which can be subscribed by any node to see the changed parameter. Note that the **/parameter_events** is a topic for all nodes on the network, and we have to rely on executors to process those.
```

This can be updated throughout the document.

 The validation callback is often abused to trigger side-effects in the code, for instance updating class attributes even before parameter has been set successfully. Instead of relying on **/parameter_events** it would be good to support an internal node interface API for registering callbacks for successful parameter set ```add_post_set_parameters_callback```.

>  We can use the proposed ```add_post_set_parameters_callback``` for setting some param but this might result in unpredictable behaviour because of infinite recusion and due to the params being set atomically. In order to get around this we propose adding another registered callback ```add_pre_set_parameters_callback``` which will be triggered before the validation callbacks and can be used to modify the parameter list as required.

Reviewer: This paragraph seems like it belongs in the next section instead of the intro, as it is describing the design.

# Proposed node parameters callback Design

## Introduction:

> The original requirement came in **gazeb_ros_pkgs** for setting individual wheel slip params based on global wheel slip value [link to original issue](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1365). 

Reviewer: ```suggestion
The original requirement came in **gazebo_ros_pkgs** for setting individual wheel slip parameters based on global wheel slip value [link to original issue](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1365). 
```

> The main idea was to set some parameters using `set_parameter` API once a given param was set successfully. 

Reviewer: ```suggestion
The main requirement is to set one or more parameters after another parameter is set successfully. 
```

> The requirement let to the discussion of supporting registered callbacks once the parameters are set and have been requested by users before.

Reviewer: ```suggestion
Additionally, it would be nice if users could be notified locally (via a callback) when parameters have been set successfully (i.e. post validation).
```

> In the current Node API, the `add_on_set_parameters_callback` is used for doing any validation for parameters before setting the parameters successfully and once the parameters are validated `ParameterEventHandler` object is used to publish any 

Reviewer: ```suggestion
With the current parameters API, the `add_on_set_parameters_callback` is intended for validation of parameter values before they are set, it should **not** cause any side-effects.
```

> changes to node parameters on **/parameter_events** topic, which can be subscribed by any node to see the changed parameter(Note that the **/parameter_events** is a topic for all nodes on the network, and we have to rely on executors to process those).

Reviewer: ```suggestion
There is also the `ParameterEventHandler` that publishes changes to node parameters on `/parameter_events` topic for external nodes to see.
Though the node could subscribe to the `/parameter_events` topic to be notified of changes to its own parameters, it is less than ideal since there is a delay caused by waiting for an executor to process the callback.
```

> We propose adding a `PostSetParametersCallbackHandle` for successful parameter set similar to OnSetParametersCallbackHandle for parameter validation, 

Reviewer: ```suggestion
We propose adding a `PostSetParametersCallbackHandle` for successful parameter set similar to `OnSetParametersCallbackHandle` for parameter validation.
```

> additionally we propose adding a `PreSetParametersCallbackHandle` which can be used to modify the parameters being set

Reviewer: ```suggestion
Additionally, we propose adding a `PreSetParametersCallbackHandle` useful for modifying the list of parameters being set.
```

> Some related discussion here [#609](https://github.com/ros2/rclcpp/issues/609) [#1789](https://github.com/ros2/rclcpp/pull/1789).

Reviewer: ```suggestion
Related discussion can be found in [#609](https://github.com/ros2/rclcpp/issues/609) and [#1789](https://github.com/ros2/rclcpp/pull/1789).
```

## Proposed Architecture

> The validation callback is often abused to trigger side effects in the code, for instance updating class attributes even before parameter has been set successfully. 

Reviewer: ```suggestion
The validation callback is often abused to trigger side effects in the code, for instance updating class attributes even before a parameter has been set successfully. 
```

> Instead of relying on **/parameter_events** it would be good to support an internal node interface API for registering callbacks for successful parameter set 
> `add_post_set_parameters_callback`.

Reviewer: ```suggestion
Instead of relying on the `/parameter_events` topic to be notified of parameter changes, users can register a callback with a new API, `add_post_set_parameters_callback`.
```

> We can use the proposed `add_post_set_parameters_callback` for setting some param but this might result in unpredictable behaviour because of infinite recursion 

Reviewer: ```suggestion
It is possible to use the proposed `add_post_set_parameters_callback` for setting additional parameters, but this might result in infinite recursion and does not allow those additional parameters to be set atomically with the original parameter(s) changed.
```

> and due to the params being set atomically. In order to get around this we propose adding another registered callback `add_pre_set_parameters_callback` which 
> will be triggered before the validation callbacks and can be used to modify the parameter list as required.

Reviewer: ```suggestion
To workaround these issues, we propose adding a "pre set" callback type that can be registered with `add_pre_set_parameters_callback`, which will be triggered before the validation callbacks and can be used to modify the parameter list.
```

![Desgin API](https://github.com/ros2/rclcpp/blob/deepanshu/local-param-changed-callback-support/rclcpp/doc/param_callback_design.png?raw=true)

## Alternatives

> * The **/parameter_events** topic can be used to monitor any changes to node parameters and use ```set_parameter``` API if required to set any further parameters. However, it seems weird that we have to rely on network to notify nodes of their own parameter changes. Therefore, it makes sense to support a callback which gets triggered after the parameter is changed, so that it can trigger some more events.

Reviewer: ```suggestion
* Users could call `set_parameter` while processing a message from the `/parameter_events` topic, However, there is extra overhead in having to create subscription (as noted earlier).
```

> * The ```add_on_set_parameters_callback``` can be used to ```set_paramters```, however the ```ParameterMutationRecursionGuard``` will not allow this since the ```add_on_set_parameters_callback``` is supposed to be used for validation purposes only.

Reviewer: I don't think the recursion guard is a strong reason to avoid this approach, since it is in our control to remove or patch the recursion guard to allow for some recursion depth before failing.
```suggestion
* Users could call `set_parameter` inside the "on set" parameters callback, however it is not well-defined how side-effects should handle cases where parameter validation fails.
```

> * The ```add_post_set_parameters_callback```  can be used to set parameters but this might result in different behaviours depending on whether the params were set atomically or not. Also, we might enter an infinite recursion depending on how params are being set. 

Reviewer: Following my previous comment, I wouldn't rule out this use-case. It's possible we could relax the recursion guard in the future.
```suggestion
```
